### PR TITLE
8321616: Retire binary test vectors in test/jdk/java/util/zip/ZipFile

### PR DESCRIPTION
**Pull request by eirbjo**
This PR suggests we retire the binary test vectors `ZipFile/input.zip`, `ZipFile/input.jar` and `ZipFile/crash.jar`

Binary test vectors are harder to analyze,  and sharing test vectors across unrelated tests increases maintenance burden. It would be better to have each test produce its own test vectors independently.

While visiting these dusty tests, we should take the opportunity to convert them to JUnit, add more comments and perform some mild modernization and cleanups where appropriate. We should also consider whether any test are duplicated and can be retired or moved into other test files as separate methods. See comments below.

To help reviewers, here are some comments on the updated tests:

`Available.java`
This test currently has no jtreg `@test` header, so isn't currently active. After discussion, we decided to merge this test into `ReadZip.java`. I added some checks to verify that reading from the stream reduces the number of available bytes accordingly, also checking the behavior when the stream is closed.

`CopyJar.java`
The concern of copying entries seems to now have better coverage in the test `zip/CopyZipFile`. We decided to retire this test rather than convert it and instead convert `CopyZipFile` to JUnit.

`EnumAfterClose.java`
To prevent confusion with Java Enums, I suggest we rename this test to `EnumerateAfterClose`.

`FinalizeInflater.java`  
The code verified by this test has been updated to use cleaners instead of finalizers. Still, this test code relies on finalizers. Not sure if this is an issue, but this test will need to be updated when finalizers are finally removed.

`GetDirEntry.java`
We decided to merge this test into `ReadZip.readDirectoryEntries` rather than keeping it as a separate test.

`ReadZip.java`
Nothing exciting here, the single main method was split into multiple JUnit methods, each focusing on a separate concern. A new test case `noentries()` was added, resolving [JDK-8322830](https://bugs.openjdk.org/browse/JDK-8322830)

`ReleaseInflater.java`
Nothing exciting, tried to add some comment to help understanding of what is tested.

`StreamZipEntriesTest.java`
This test was using TestNG so was converted to JUnit for consistency. Added some comments to help understanding.

`ReadAfterClose.java`
This uses the binary test vector `crash.jar`. The test is converted to JUnit and moved to `ReadZip.readAfterClose´. The test is expanded to exercise more read methods.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8321616](https://bugs.openjdk.org/browse/JDK-8321616): Retire binary test vectors in test/jdk/java/util/zip/ZipFile (**Enhancement** - P4)
 * [JDK-8322830](https://bugs.openjdk.org/browse/JDK-8322830): Add test case for ZipFile opening a ZIP with no entries (**Enhancement** - P4)

### Reviewers
 * [Lance Andersen](https://openjdk.org/census#lancea) (@LanceAndersen - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17038/head:pull/17038` \
`$ git checkout pull/17038`

Update a local copy of the PR: \
`$ git checkout pull/17038` \
`$ git pull https://git.openjdk.org/jdk.git pull/17038/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17038`

View PR using the GUI difftool: \
`$ git pr show -t 17038`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17038.diff">https://git.openjdk.org/jdk/pull/17038.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17038#issuecomment-1850278566)

### CODE REVIEW

**Comment by LanceAndersen** [comment]
Thank you for tacking the conversion of this test to junit

One quick comment, if we are updating this test, we should look to get rid of input.zip which means we should also address as part of the change:

open/test/jdk/java/util/zip/ZipFile/EnumAfterClose.java
open/test/jdk/java/util/zip/ZipFile/FinalizeInflater.java
open/test/jdk/java/util/zip/ZipFile/StreamZipEntriesTest.java

**Comment by eirbjo** [comment]
> One quick comment, if we are updating this test, we should look to get rid of input.zip 

I started going down that road, but felt uneasy about the amount of unrelated changes in a single PR. I'd like to make efficient use of reviewer time, so my preference was to focus on the JUnit conversion, without too many other changes.

If you prefer that we get rid of input.zip and also convert the affected tests to JUnit in the same PR, then I'm happy to switch strategy.

Note that `StreamZipEntriesTest` also uses `input.jar`, this is also read by `Available`, `CopyJar`, `GetDirEntry`, `ReleaseInflater`. Should we get rid of `input.jar` as well, or perhaps defer that to a follow-up?

**Comment by eirbjo** [comment]
The scope of this PR has now expanded to removing uses of the `input.zip` and `input.jar` files, updating any test using them to produce their own test vectors, and convert affected tests to JUnit.

I'm marking the PR ready for review again. Before looking too closely at the code, it would be useful to discuss the following tests:

- `Available.java`: This test has no jtreg header. I've added one and converted the test. Is this worthwhile, or should we rather remove it?
- `CopyJar.java`: The concern tested seems to have superior coverage in the test `zip/CopyZipFile.java`. Should we retire `CopyJar.java` instead of coverting it?
- `DirEntry.java`: There is duplication between this test and `ReadZip. readDirectoryEntry`. Should we retire one of these?

**Comment by LanceAndersen** [comment]
> The scope of this PR has now expanded to removing uses of the `input.zip` and `input.jar` files, updating any test using them to produce their own test vectors, and convert affected tests to JUnit.
> 
> I'm marking the PR ready for review again. Before looking too closely at the code, it would be useful to discuss the following tests:
> 
> * `Available.java`: This test has no jtreg header. I've added one and converted the test. Is this worthwhile, or should we rather remove it?

This could be moved into ReadZip.  I do not believe we have a specific test and it is trivial

> * `CopyJar.java`: The concern tested seems to have superior coverage in the test `zip/CopyZipFile.java`. Should we retire `CopyJar.java` instead of coverting it?

Yes CopyZipFile already exercises Zipfile.ZipInputStream so it is safe to retire CopyJar (though CopyZipFile could use a junit conversion ;-) 
> * `DirEntry.java`: There is duplication between this test and `ReadZip.readDirectoryEntry()`. Should we retire one of these?

I believe you meant GetDirEntry.java not DirEntry.java?

Having a test that specifically validates we can read META-INF is not a bad thing, but I suspect we have a test that already does that if not in the java/util/zip tests or java/util/jar tests.  If not we should keep it but merge it as you suggest

**Comment by eirbjo** [comment]
@LanceAndersen 

Thanks for your guidance! I moved `Available` into `ReadZip`, deleted `CopyJar` and merged `GetDirEntry` into ReadZip.readDirectoryEntries (adding a 'META-INF/' directory just in case)

**Comment by eirbjo** [comment]
> (though CopyZipFile could use a junit conversion ;-)

Agreed, I have included that conversion in this PR :-) This test can make good use of `assertThrows` and splitting different concerns into separate test methods.

**Comment by eirbjo** [comment]
Seeing that `ReadAfterClose.java` uses a binary test vector `crash.jar`, I think it makes sense to include it in this PR, convert it to JUnit and move it into `ReadZip.readAfterClose`.

This removes the last remaining binary test vector ZIP in the `ZipFile/` directory.

**Comment by LanceAndersen** [comment]
> Seeing that `ReadAfterClose.java` uses a binary test vector `crash.jar`, I think it makes sense to include it in this PR, convert it to JUnit and move it into `ReadZip.readAfterClose`.
> 
> This removes the last remaining binary test vector ZIP in the `ZipFile/` directory.

These are on my list to review, thank you for the update.

One comment, for the tests that we are removing/retiring, we probably want to include the bug numbers and/or the name of the original test class in the new home for the tests to  make it easier for future maintainers to chase back the history in the unlikely event that an issue arises...

**Review by LanceAndersen — APPROVED** [review]
Changes look fine.  Probably need to change the copyright from 2023 -> 2024

File: test/jdk/java/util/zip/ZipFile/ReadZip.java
Comment (eirbjo) [reply test/jdk/java/util/zip/ZipFile/ReadZip.java]: Fixed.

**Comment by eirbjo** [comment]
Thanks for your review, @LanceAndersen!

You might want to take a quick look at the lastest updates addressing your review before I integrate.

**Comment by LanceAndersen** [comment]
> Thanks for your review, @LanceAndersen!
> 
> You might want to take a quick look at the lastest updates addressing your review before I integrate.

Looks good.  I am running mach5 tiers1-3 internally once it completes, I will approve again

**Review by LanceAndersen — APPROVED** [review]
Internal mach5 tiers 1-3 completed

**Pull request closed without merging**
